File: Go/pkg/transform/conversion.go
package transform

/*
	Package 'transform' provides mechanism's for converting data between different formats.
*/

import (
	"fmt"
	"log/slog"
	"strconv"
	"strings"
)

/*
	The Conversion module provides methods to convert data between hexadecimal and ASCII formats,
	according to the configured transmission mode.
*/

const (
	TransmissionModeASCII = "ascii"
	TransmissionModeHex   = "hex"
)

type Conversion struct {
	mode string
}

func NewConversion(mode string) *Conversion {
	return &Conversion{
		mode: mode,
	}
}

// FromHexToAscii converts a hexadecimal message to its ASCII representation.
func (c *Conversion) FromHexToAscii(message []byte, prefix string, suffix string) []byte {
	var msg []byte
	length := len(message)
	for i := range length {
		byteString := fmt.Sprintf("%02X", message[i])
		if c.isAscii(byteString[0]) && c.isAscii(byteString[1]) {
			msg = append(msg, byteString[0:1][0])
			msg = append(msg, byteString[1:2][0])
		} else if c.isAllowed(message[i], prefix, suffix, i, length) {
			msg = append(msg, message[i])
		}
	}
	return msg
}

// FromAsciiToHex converts an ASCII message to its hexadecimal representation.
func (c *Conversion) FromAsciiToHex(message []byte, prefix string, suffix string) []byte {
	var msg []byte
	length := len(message)
	i := 0
	for i < length {
		if c.isAscii(message[i]) {
			if i+1 < length && c.isAscii(message[i+1]) {
				hexStr := string(message[i]) + string(message[i+1])
				hexVal, err := strconv.ParseUint(hexStr, 16, 8)
				if err == nil {
					msg = append(msg, byte(hexVal))
					i += 2
				}
			}
		} else if c.isAllowed(message[i], prefix, suffix, i, length) {
			msg = append(msg, message[i])
			i++
		} else {
			slog.Error("Invalid ASCII character in message: " + string(message[i]))
			i++
		}
	}

	return msg
}

// Helper methods
// isAscii checks if a byte represents a valid ASCII hexadecimal character (0-9, A-F, a-f).
func (c *Conversion) isAscii(b byte) bool {
	if b >= 0x30 && b <= 0x39 {
		return true
	}
	if b >= 0x41 && b <= 0x46 {
		return true
	}
	if b >= 0x61 && b <= 0x66 {
		return true
	}
	return false
}

// isAllowed checks if a byte is allowed in the conversion based on prefix or suffix, considering the position in the message.
func (c *Conversion) isAllowed(b byte, prefix string, suffix string, index int, length int) bool {
	if strings.Contains(prefix, string(b)) && index < len(prefix) {
		return true
	}
	if strings.Contains(suffix, string(b)) && index >= length-len(suffix) {
		return true
	}
	slog.Warn("Byte " + fmt.Sprintf("%02X", b) + " is not allowed in conversion at index " + fmt.Sprintf("%d", index))
	return false
}

// Display returns a string representation of the message in hex format for display purposes.
func (c *Conversion) Display(msg []byte, prefix string, suffix string) string {
	display := msg
	if c.mode == TransmissionModeASCII {
		display = c.FromAsciiToHex(msg, prefix, suffix)
	}
	var displayMsg string
	for _, b := range display {
		displayMsg += fmt.Sprintf("%02X", b)
	}
	return displayMsg
}

// GetHexMessage retrieves the message in hexadecimal format.
func (c *Conversion) GetHexMessage(msg []byte, prefix string, suffix string) []byte {

	if c.mode == TransmissionModeASCII {
		return c.FromAsciiToHex(msg, prefix, suffix)
	}
	return msg
}

// GetConvertedMessage retreives the message based on the conversion mode.
func (c *Conversion) GetConvertedMessage(msg []byte, prefix string, suffix string) []byte {
	if c.mode == TransmissionModeASCII {
		return c.FromHexToAscii(msg, prefix, suffix)
	}
	return msg
}

File: Python/pkg/transform/conversion.py

"""
	Package 'transform' provides mechanism's for converting data between different formats.
"""

from constants import Mode

"""
    The Conversion class provides methods to convert messages between hexadecimal and ASCII formats,
    according to the configured transmission mode.
"""

class Conversion:
     
    def __init__(self, mode: str, logger):
        self.logger = logger
        self.mode = mode
          
    def from_hex_to_ascii(self, message: bytearray, prefix: str, suffix: str):
        """
        Converts a hexadecimal message to its ASCII representation.

        Args:
            message (bytearray): The message bytearray.
            prefix (str): The prefix string.
            suffix (str): The suffix string.
        Returns:
            bytearray: The converted ASCII bytearray.
        Raises:
            None
        """
        if self.mode != Mode.ASCII.value:
            return message
        msg = bytearray()
        length = len(message)
        for index, byte in enumerate(message):
            byte_string = f'{byte:02X}'
            if self.is_ascii(ord(byte_string[0])) and self.is_ascii(ord(byte_string[1])):
                msg.append(ord(byte_string[0]))
                msg.append(ord(byte_string[1]))
            elif self.is_allowed_ascii(byte, prefix, suffix, index, length):
                msg.append(byte)                
        return msg
   
    def from_ascii_to_hex(self, message: bytearray, prefix: str, suffix: str):
        """
        Converts an ASCII message to its hexadecimal representation.

        Args:
            message (bytearray): The ASCII message bytearray.
            prefix (str): The prefix string.
            suffix (str): The suffix string.
        Returns:
            bytearray: The converted hexadecimal bytearray.
        Raises:
            ValueError: If the message contains invalid ASCII characters.
        """
        if self.mode != Mode.ASCII.value:
            return message
        msg = bytearray()
        length = len(message)
        i = 0
        while i < length:
            if self.is_ascii(message[i]):
                hex_str = chr(message[i]) + chr(message[i+1])
                byte_val = int(hex_str, 16)
                msg.append(byte_val)
                i += 2
            elif self.is_allowed_ascii(message[i], prefix, suffix, i, length):
                msg.append(message[i])
                i += 1
            else:
                raise ValueError("Invalid ASCII character in message: %02.X", message[i])
        return msg
    
    # Helper Methods
    def is_ascii(self, char: int):
        """
        Checks if a byte represents a valid ASCII hexadecimal charater (0-9, A-F, a-f).

        Args:
            char (int): The character to check.
        Returns:
            bool: True if the character is a valid ASCII hexadecimal character (0-9, A-F, a-f), 
            False otherwise.
        Raises:
            None
        """
        if (char >= 0x30 and char <= 0x39):
            return True
        if (char >= 0x41 and char <= 0x46):
            return True
        if (char >= 0x61 and char <= 0x66):
            return True
        return False
    
    def is_allowed_ascii(self, char: int, prefix: str, suffix: str, index: int, length: int):
        """
        Checks if a byte is allowed in the conversion based on prefix or suffix, 
        considering the position in the message.

        Args:
            char (int): The character to check.
        Returns:
            bool: True if the character is allowed ASCII, False otherwise.
        Raises:
            None
        """
        if chr(char) in prefix and index < len(prefix):
            return True        
        if chr(char) in suffix and index >= length - len(suffix):
            return True
        self.logger.warning("Byte: %02.X is not allowed in conversion at index %d", char, index)
        return False
    
    def display(self, msg: bytearray):
        """
        Returns a string representation of the message in hex format for display purposes.

        Args:
            msg (bytearray): The message bytearray to display.
        Returns:
            string: The message in the hex format.
        Raises:
            None
        """
        if self.mode == Mode.ASCII.value:
            disp = msg.decode('utf-8')
        else:
            disp = msg.hex().upper()
        return disp
    
    def get_hex_message(self, msg: bytearray, prefix: str, suffix: str):
        """
        Retrieves the message in hexadecimal format.

        Args:
            msg (bytearray): The message bytearray.
            prefix (str): The prefix string.
            suffix (str): The suffix string.
        Returns:
            bytearray: The message in hexadecimal format.
        Raises:
            None
        """
        if self.mode == Mode.ASCII.value:
            return self.from_ascii_to_hex(msg, prefix, suffix)
        else:
            return msg
    
    def get_converted_message(self, msg: bytearray, prefix: str, suffix: str):
        """
        Retrieves the message based on the conversion mode.

        Args:
            msg (bytearray): The message bytearray.
            prefix (str): The prefix string.
            suffix (str): The suffix string.
        Returns:
            bytearray: The converted message in the appropriate format.
        Raises:
            None
        """
        if self.mode == Mode.ASCII.value:
            return self.from_hex_to_ascii(msg, prefix, suffix)
        else:
            return msg
